### ISSUE TITLE
jOOQ assumes _rowid_ is present in SQLite WITHOUT ROWID tables

### ISSUE THREAD

**Issue by JacksonBailey** [jOOQ/jOOQ#14656]
### Expected behavior

jOOQ should not try to use the `_rowid_` column on SQLite tables using the `WITHOUT ROWID` table option because it does not exist. *(I am tempted to say it should never try to use `_rowid_` directly and let SQLite handle that transparently but I'm sure I'm missing a lot of important context the jOOQ team has here.)*

To save you a step, this is SQLite's documentation for [Clustered Indexes and the WITHOUT ROWID Optimization](https://www.sqlite.org/withoutrowid.html) and [ROWIDs and the INTEGER PRIMARY KEY](https://www.sqlite.org/lang_createtable.html#rowid).

### Actual behavior

jOOQ tries to access the `_rowid_` column which causes an error because no such column exists in a `WITHOUT ROWID` table.

Related to #9711.

### Summary

Generate code against a SQLite database table using the `WITHOUT ROWID` option and then try to `store()` a row using the generated `org.jooq.Record` from `create.newRecord(Tables.YOUR_TABLE_HERE)`.

To reproduce you can run `./gradlew check` on [my code](https://github.com/JacksonBailey/wheel/tree/sqlite-rowid-jooq-bug) if you'd like, but I will share all relevant logs. The Gradle output is [here](https://github.com/JacksonBailey/wheel/actions/runs/4212580977/jobs/7311704048) but it does not show a useful stack trace of the test failure, but I will include that below.

### DDL

```sql
CREATE TABLE flag (
    name TEXT PRIMARY KEY NOT NULL,
    state INTEGER NOT NULL CONSTRAINT state_is_boolean CHECK (state IN (0, 1))
) WITHOUT ROWID;
```

### Codegen output

jOOQ *does* correctly error when encountering the `WITHOUT ROWID` table option which is good because it cannot handle such tables.

```
14:01:18 INFO Initialising properties  : C:\Users\Jackson\Code\wheel\vexillum\vexillum-sqlite\build\tmp\generateJooq\config.xml
14:01:19 INFO

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@  @@        @@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@        @@@@@@@@@@
@@@@@@@@@@@@@@@@  @@  @@    @@@@@@@@@@
@@@@@@@@@@  @@@@  @@  @@    @@@@@@@@@@
@@@@@@@@@@        @@        @@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@        @@        @@@@@@@@@@
@@@@@@@@@@    @@  @@  @@@@  @@@@@@@@@@
@@@@@@@@@@    @@  @@  @@@@  @@@@@@@@@@
@@@@@@@@@@        @@  @  @  @@@@@@@@@@
@@@@@@@@@@        @@        @@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@  @@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@  Thank you for using jOOQ 3.17.8

14:01:19 INFO

jOOQ tip of the day: jOOQ offers out of the box functionality for common scenarios related to audit columns: https://www.jooq.org/doc/dev/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-audit/

14:01:19 INFO No <inputCatalog/> was provided. Generating ALL available catalogs instead.
14:01:19 INFO No <inputSchema/> was provided. Generating ALL available schemata instead.
14:01:20 INFO License parameters
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO   Thank you for using jOOQ and jOOQ's code generator
14:01:20 INFO
14:01:20 INFO Database parameters
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO   dialect                : SQLITE
14:01:20 INFO   URL                    : jdbc:sqlite:C:\Users\Jackson\Code\wheel\vexillum\vexillum-sqlite\build\generated\db\vexillum.sqlite
14:01:20 INFO   target dir             : C:\Users\Jackson\Code\wheel\vexillum\vexillum-sqlite\build\generated-src\jooq\main
14:01:20 INFO   target package         : dev.jacksonbailey.vexillum.db.jooq
14:01:20 INFO   includes               : []
14:01:20 INFO   excludes               : [flyway_schema_history]
14:01:20 INFO   includeExcludeColumns  : false
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO
14:01:20 INFO JavaGenerator parameters
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO annotations
14:01:20 INFO   generated              : true
14:01:20 INFO   JPA                    : true
14:01:20 INFO   JPA version            :
14:01:20 INFO   validation             : true
14:01:20 INFO   spring                 : false
14:01:20 INFO comments
14:01:20 INFO   comments               : true
14:01:20 INFO   on attributes          : true
14:01:20 INFO   on catalogs            : true
14:01:20 INFO   on columns             : true
14:01:20 INFO   on embeddables         : true
14:01:20 INFO   on keys                : true
14:01:20 INFO   on links               : true
14:01:20 INFO   on packages            : true
14:01:20 INFO   on parameters          : true
14:01:20 INFO   on queues              : true
14:01:20 INFO   on routines            : true
14:01:20 INFO   on schemas             : true
14:01:20 INFO   on sequences           : true
14:01:20 INFO   on tables              : true
14:01:20 INFO   on udts                : true
14:01:20 INFO sources
14:01:20 INFO   sources                : true
14:01:20 INFO   sources on views       : true
14:01:20 INFO global references
14:01:20 INFO   global references      : true
14:01:20 INFO   catalogs               : true
14:01:20 INFO   domains                : true
14:01:20 INFO   indexes                : true
14:01:20 INFO   keys                   : true
14:01:20 INFO   links                  : true
14:01:20 INFO   queues                 : true
14:01:20 INFO   routines               : true
14:01:20 INFO   schemas                : true
14:01:20 INFO   sequences              : true
14:01:20 INFO   tables                 : true
14:01:20 INFO   udts                   : true
14:01:20 INFO object types
14:01:20 INFO   daos                   : false
14:01:20 INFO   indexes                : true
14:01:20 INFO   instance fields        : true
14:01:20 INFO   interfaces             : false
14:01:20 INFO   interfaces (immutable) : false
14:01:20 INFO   javadoc                : true
14:01:20 INFO   keys                   : true
14:01:20 INFO   links                  : true
14:01:20 INFO   pojos                  : true
14:01:20 INFO   pojos (immutable)      : true
14:01:20 INFO   queues                 : true
14:01:20 INFO   records                : true
14:01:20 INFO   routines               : true
14:01:20 INFO   sequences              : true
14:01:20 INFO   sequenceFlags          : true
14:01:20 INFO   table-valued functions : true
14:01:20 INFO   tables                 : true
14:01:20 INFO   udts                   : true
14:01:20 INFO   relations              : true
14:01:20 INFO other
14:01:20 INFO   deprecated code        : true
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO
14:01:20 INFO Generation remarks
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO Generating catalogs      : Total: 1
14:01:20 INFO ARRAYs fetched           : 0 (0 included, 0 excluded)
14:01:20 INFO Domains fetched          : 0 (0 included, 0 excluded)
14:01:20 INFO Tables fetched           : 2 (1 included, 1 excluded)
14:01:20 INFO Embeddables fetched      : 0 (0 included, 0 excluded)
14:01:20 INFO Enums fetched            : 0 (0 included, 0 excluded)
14:01:20 INFO Packages fetched         : 0 (0 included, 0 excluded)
14:01:20 INFO Routines fetched         : 0 (0 included, 0 excluded)
14:01:20 INFO Sequences fetched        : 0 (0 included, 0 excluded)
14:01:20 INFO No schema version is applied for catalog . Regenerating.
14:01:20 INFO
14:01:20 INFO Generating catalog       : DefaultCatalog.java
14:01:20 INFO ==========================================================
14:01:20 INFO Generating schemata      : Total: 1
14:01:20 INFO No schema version is applied for schema . Regenerating.
14:01:20 INFO Generating schema        : DefaultSchema.java
14:01:20 INFO ----------------------------------------------------------
14:01:20 INFO UDTs fetched             : 0 (0 included, 0 excluded)
14:01:20 INFO Generating tables
14:01:20 INFO Cannot parse SQL: CREATE TABLE flag (
    name TEXT PRIMARY KEY NOT NULL,
    state INTEGER NOT NULL CONSTRAINT state_is_boolean CHECK (state IN (0, 1))
) WITHOUT ROWID
org.jooq.impl.ParserException: Unexpected clause: [4:3] ...RAINT state_is_boolean CHECK (state IN (0, 1))
) [*]WITHOUT ROWID
        at org.jooq.impl.DefaultParseContext.exception(ParserImpl.java:14146)
        at org.jooq.impl.DefaultParseContext.done(ParserImpl.java:14332)
        at org.jooq.impl.DefaultParseContext.lambda$parseQuery0$2(ParserImpl.java:936)
        at org.jooq.impl.DefaultParseContext.wrap(ParserImpl.java:14339)
        at org.jooq.impl.DefaultParseContext.parseQuery0(ParserImpl.java:936)
        at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:753)
        at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:748)
        at org.jooq.meta.sqlite.SQLiteDatabase.loadCheckConstraints(SQLiteDatabase.java:333)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations0$27(AbstractDatabase.java:3119)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3753)
        at org.jooq.meta.AbstractDatabase.getRelations0(AbstractDatabase.java:3119)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations$18(AbstractDatabase.java:2791)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3753)
        at org.jooq.meta.AbstractDatabase.getRelations(AbstractDatabase.java:2791)
        at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey(DefaultColumnDefinition.java:154)
        at org.jooq.meta.AbstractTableDefinition.getPrimaryKey(AbstractTableDefinition.java:96)
        at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:5873)
        at org.jooq.codegen.JavaGenerator.generateTables(JavaGenerator.java:5852)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:652)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:607)
        at org.jooq.codegen.JavaGenerator.generate0(JavaGenerator.java:519)
        at org.jooq.codegen.AbstractGenerator.generate(AbstractGenerator.java:187)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:227)
        at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:942)
        at org.jooq.codegen.GenerationTool.run(GenerationTool.java:240)
        at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:235)
        at org.jooq.codegen.GenerationTool.main(GenerationTool.java:207)

14:01:20 INFO Generating table         : Flag.java [input=flag, pk=pk_flag]
14:01:20 INFO Indexes fetched          : 0 (0 included, 0 excluded)
14:01:20 INFO Tables generated         : Total: 547.589ms
14:01:20 INFO Generating table POJOs
14:01:20 INFO Generating POJO          : Flag.java
14:01:20 INFO Table POJOs generated    : Total: 555.404ms, +7.814ms
14:01:20 INFO Generating table references
14:01:20 INFO Table refs generated     : Total: 559.507ms, +4.102ms
14:01:20 INFO Generating Keys
14:01:20 INFO Keys generated           : Total: 568.48ms, +8.972ms
14:01:20 INFO Generating Indexes
14:01:20 INFO Skipping empty indexes
14:01:20 INFO Generating table records
14:01:20 INFO Generating record        : FlagRecord.java
14:01:20 INFO Table records generated  : Total: 618.982ms, +50.502ms
14:01:20 INFO Generation finished:     : Total: 619.901ms, +0.919ms
14:01:20 INFO
14:01:20 INFO Affected files: 7
14:01:20 INFO Modified files: 7
14:01:20 INFO Removing excess files
```

### Failing unit test

I've marked what line 49 is because it is mentioned in the stack trace.

```
  @Test
  void test() {
    assertEquals(0, create.fetchCount(Tables.FLAG));
    var newFlag = create.newRecord(Tables.FLAG);
    newFlag.setName("WOW");
    newFlag.setState(1);
    newFlag.store(); // This is line 49
    assertEquals(1, create.fetchCount(Tables.FLAG));
    var result = create.fetchOne(Tables.FLAG, Tables.FLAG.NAME.eq("WOW"));
    assertEquals(1, result.getState());
  }
```

### Test stack trace

```
org.jooq.exception.DataAccessException: SQL [select flag.name from flag where _rowid_ = last_insert_rowid()]; [SQLITE_ERROR] SQL error or missing database (no such column: _rowid_)
	at org.jooq_3.17.8.SQLITE.debug(Unknown Source)
	at org.jooq_3.17.8.SQLITE.debug(Unknown Source)
	at app//org.jooq.impl.Tools.translate(Tools.java:3313)
	at app//org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:678)
	at app//org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:355)
	at app//org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:290)
	at app//org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2837)
	at app//org.jooq.impl.AbstractDMLQuery.execute(AbstractDMLQuery.java:1044)
	at app//org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:341)
	at app//org.jooq.impl.TableRecordImpl.storeInsert0(TableRecordImpl.java:192)
	at app//org.jooq.impl.TableRecordImpl.lambda$storeInsert$0(TableRecordImpl.java:158)
	at app//org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144)
	at app//org.jooq.impl.TableRecordImpl.storeInsert(TableRecordImpl.java:157)
	at app//org.jooq.impl.UpdatableRecordImpl.store0(UpdatableRecordImpl.java:222)
	at app//org.jooq.impl.UpdatableRecordImpl.lambda$store$0(UpdatableRecordImpl.java:149)
	at app//org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144)
	at app//org.jooq.impl.UpdatableRecordImpl.store(UpdatableRecordImpl.java:148)
	at app//org.jooq.impl.UpdatableRecordImpl.store(UpdatableRecordImpl.java:140)
	at app//dev.jacksonbailey.wheel.vexillum.db.SqliteFlagStoreTest.test(SqliteFlagStoreTest.java:49)
	at java.base@19.0.2/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base@19.0.2/java.lang.reflect.Method.invoke(Method.java:578)
	at app//org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at app//org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base@19.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base@19.0.2/java.util.ArrayList.forEach(ArrayList.java:1511)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at app//org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:110)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:90)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:85)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
	at java.base@19.0.2/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base@19.0.2/java.lang.reflect.Method.invoke(Method.java:578)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (no such column: _rowid_)
	at app//org.sqlite.core.DB.newSQLException(DB.java:1179)
	at app//org.sqlite.core.DB.newSQLException(DB.java:1190)
	at app//org.sqlite.core.DB.throwex(DB.java:1150)
	at app//org.sqlite.core.NativeDB.prepare_utf8(Native Method)
	at app//org.sqlite.core.NativeDB.prepare(NativeDB.java:126)
	at app//org.sqlite.core.DB.prepare(DB.java:264)
	at app//org.sqlite.core.CorePreparedStatement.<init>(CorePreparedStatement.java:46)
	at app//org.sqlite.jdbc3.JDBC3PreparedStatement.<init>(JDBC3PreparedStatement.java:31)
	at app//org.sqlite.jdbc4.JDBC4PreparedStatement.<init>(JDBC4PreparedStatement.java:25)
	at app//org.sqlite.jdbc4.JDBC4Connection.prepareStatement(JDBC4Connection.java:34)
	at app//org.sqlite.jdbc3.JDBC3Connection.prepareStatement(JDBC3Connection.java:226)
	at app//org.sqlite.jdbc3.JDBC3Connection.prepareStatement(JDBC3Connection.java:206)
	at app//org.jooq.impl.ProviderEnabledConnection.prepareStatement(ProviderEnabledConnection.java:109)
	at app//org.jooq.impl.SettingsEnabledConnection.prepareStatement(SettingsEnabledConnection.java:82)
	at app//org.jooq.impl.ProviderEnabledConnection.prepareStatement(ProviderEnabledConnection.java:109)
	at app//org.jooq.impl.SettingsEnabledConnection.prepareStatement(SettingsEnabledConnection.java:82)
	at app//org.jooq.impl.AbstractResultQuery.prepare(AbstractResultQuery.java:216)
	at app//org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:308)
	... 94 more
```

### jOOQ Version

jOOQ Open Source 3.17.8

### Database product and version

SQLite 3.40.1 (bundled with org.xerial:sqlite-jdbc:3.40.1.0)

### Java Version

openjdk version "19" 2022-09-20

### OS Version

Microsoft Windows 10

### JDBC driver name and version (include name if unofficial driver)

org.xerial:sqlite-jdbc:3.40.1.0

**Comment by lukaseder** [comment]
Thanks a lot for your detailed report

> (I am tempted to say it should never try to use `_rowid_` directly and let SQLite handle that transparently but I'm sure I'm missing a lot of important context the jOOQ team has here.)

Well,

1. It's quite the documented feature: https://www.sqlite.org/autoinc.html
2. How else would you fetch generated keys, other than with `last_insert_rowid()`?

But obviously, indeed, jOOQ shouldn't attempt to fetch that on your table, because it simply doesn't make sense. There has been some historic reason to auto-fetch identities *and* primary keys (in the absence of identities), in case the identity is hidden from jOOQ (e.g. through a trigger generated sequence value).

The most reasonable fix is probably there somewhere. You can turn off automatic fetching of identities (for all `UpdatableRecord` types) using `Settings.returnIdentityOnUpdatableRecord`.

I'll think about what to do about this case. Perhaps, automatically fetching primary key values without known identity was a mistake, or should have been another settings governed feature that people can opt in to. After all, there are only few dialects which are capable of returning non-identity values immediately, and SQLite isn't one of them.

**Comment by lukaseder** [comment]
Parser support (ignoring the clause) for `WITHOUT ROWID` is being tracked here: https://github.com/jOOQ/jOOQ/issues/14690

**Comment by lukaseder** [comment]
Hah, SQLite is such a mess. So, here's how we decide whether a column is an identity in `SQLiteTableDefinition`:

```java
if (pk > 0) {

    // [#6854] sqlite_sequence only contains identity information once a table contains records.
    identity |= existsSqliteSequence() && create()
        .fetchOne("select count(*) from sqlite_sequence where name = ?", getName())
        .get(0, Boolean.class);

    if (!identity && !create().fetchExists(selectOne().from("{0}", DSL.name(getName()))))
        identity = getSource().matches("(?s:.*\\b" + getName() + "\\b[^,]*(?i:\\bautoincrement\\b)[^,]*.*)");
}
```

1. First, we check if there's a primary key
2. Then we check if the `sqlite_sequence` contains a value for the table (by convention, if sequence name = table name, then it's the identity, there are no other user sequences supported at the moment)
3. But because that table contains sequence values only once the first row has been inserted (see #6854), we also check if there's an autoincrement marker.
4. But actually, that marker is irrelevant. We can create a table with a sequence like below:

```sql
create table x (
  id integer not null primary key,
  val int
);
insert into x (val) values (1);
insert into x (val) values (1);
insert into x (val) values (1);
select * from x;
```

That produces:

```
|id |val|
|---|---|
|1  |1  |
|2  |1  |
|3  |1  |
```

One might be tempted to assume that we should check for the inverse, namely for the presence of `WITHOUT ROWID`. But that's not possible either, `WITHOUT ROWID` is implicit (but not listed in `SQLITE_MASTER`), if we do this:

```sql
create table x (
  id int not null primary key,
  val int
);
```

Observe the subtle difference between using `int` and `integer`. This statement now fails:

```sql
insert into x (val) values (1);
```

> SQL Error [19]: [SQLITE_CONSTRAINT_NOTNULL] A NOT NULL constraint failed (NOT NULL constraint failed: x.id)

Turns out, `INTEGER` is a `BIGINT`, which means `ROWID` (because, duh). This is documented here:
https://www.sqlite.org/lang_createtable.html#rowid

And we already have two issues related to misunderstanding all of these subtleties:

- https://github.com/jOOQ/jOOQ/issues/9020
- https://github.com/jOOQ/jOOQ/issues/11675

So, what we *seem to* have to do is:

- Fix the various attempts of recognising an identity in the code generator
- Stop attempting to fetch the `_rowid_` if we don't know any identity

With these measures, your issue should be fixed.

**Comment by lukaseder** [comment]
I'll just leave these here: 😅

![image](https://user-images.githubusercontent.com/734593/220935964-f4adb298-0ca2-44e2-a26e-5dd2e470a815.png)

![image](https://user-images.githubusercontent.com/734593/220935990-976dbfaf-9461-4dc2-8e6f-0bb506d18805.png)

![image](https://user-images.githubusercontent.com/734593/220936027-218ccc3c-c807-43bd-b47d-bc90cac32a9c.png)

![image](https://user-images.githubusercontent.com/734593/220936056-f5b07268-2e9f-4faf-8423-505f23044bc1.png)

Not sure if I should really try to formally reverse engineer all of these cases...

**Comment by lukaseder** [comment]
Naturally, none of those cases produce an entry in `sqlite_sequence`, so if we want to support these, it seems we really have to go all in on reverse engineering.

**Comment by lukaseder** [comment]
It gets better!

![image](https://user-images.githubusercontent.com/734593/220938035-5003a517-fd1a-437d-bc1c-ab8d56cc1f68.png)

**Comment by lukaseder** [comment]
The problem is really, there doesn't seem to be any reliable way to detect whether SQLite thinks a column is an identity or not, unless there are:

- Clear signs in favour (`sqlite_sequence` entry, `autoincrement` keyword present)
- Clear signs against it (`without rowid` marker in the table declaration)

In any other case, I suspect it would be safest for jOOQ to continue assuming there's a `_rowid_` on the primary key (ignoring the case where the PK is composite, containing the identity!). If your runtime issue is to be fixed, we'll have to generate the identity information on all of these tables, which will probably produce a few false positives.

As such, any fix won't be backported, this seems too brittle.

**Comment by lukaseder** [comment]
OK, I still got it wrong 😅 So, despite the fact that this `INTEGER` (not `INT`) column (which is likely, but not necessarily the identity) can be the `_rowid_`, it doesn't have to be. `_rowid_` and identity are two related but subtly independent concepts. `_rowid_` is *always* present, *unless* `WITHOUT ROWID` is supplied.

I guess (and I will probably revise this guess once more) the only way we can fix this bug is by generating some additional information about the absence of the `_rowid_` in such tables.

**Comment by lukaseder** [comment]
Maybe, this issue will become obsolete with the support of `RETURNING` in SQLite!

```sql
create table t (i integer primary key autoincrement);
insert into t default values returning *;
```

See:

- https://github.com/jOOQ/jOOQ/issues/15316

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/15316 has been implemented for jOOQ 3.19.

**Comment by lukaseder** [comment]
With proper configuration, this is now a thing of the past:

- `INSERT .. RETURNING` is now available in jOOQ for SQLite
- We're still using the `_rowid_` check in case we're trying to fetch trigger values (which aren't produced by `RETURNING`). This can be turned off:
  - By using `Settings.fetchTriggerValuesAfterReturning`
  - By using the commercial editions, which produce trigger meta information so we can omit this check

It might still be useful to generate the meta information `WITHOUT ROWID` to prevent the `_rowid_` usage in the future, but it's not a priority right now.
